### PR TITLE
[openstack] Custom dns prefixes and domains

### PR DESCRIPTION
**Pull request by bogdando**
Support separate public/private DNS records, yet limited to an                  
external DNS managed elsewhere. Users may benefit from this                     
feature by sending nsupdates to private and/or public servers                   
as they want it.                                                                
                                                                                
Additional changes:                                                             
                                                                                
* Name Heat stack by openshift_openstack_full_public_dns_domain                 
* Add openshift_openstack_public_dns_domain to the role defaults                
  to not rely on the group vars example only.                                   
* Move openshift_master_cluster_public_hostname                                 
  and openshift_master_default_subdomain from the group vars                    
  example to the role defaults, where the values these take are                 
  now evaluated.                                                                
* Evaluate openshift_master_cluster_hostname based on a private                 
  domain name as it may be different to the public name now. 

Describe configuration scenarios for external public and private
DNS servers managing public and custom private domain records for the
openshift cluster nodes.

Specify manual commands required to deploy DNS service with automation
playbooks for osp 10 on ocp 3.6 ref arch coming from openshift-ansible-contrib.

This improves UX with DNS stack not managed in openshift-ansible.

**How to test:**

* With defaults, e2e should pass
* Follow the included guide to verify an external public and private DNS server.

### CODE REVIEW

**Comment by bogdando** [comment]
@tomassedovic PTAL

**Comment by bogdando** [comment]
e2e with defaults works for me, @tomassedovic 
I'll test with a custom external DNS hosting private records as well.

**Comment by bogdando** [comment]
@tomassedovic @e-minguez PTAL . The provided documentation for external DNS services deployment works for me locally. I was using these custom vars for osp10-dns-service's [vars.yaml](https://raw.githubusercontent.com/bogdando/ocp-inventory/master/osp10-refarch-dns-service.yml) and openshift cluster's [external_dns.yaml](https://raw.githubusercontent.com/bogdando/ocp-inventory/master/external_dns.yaml).

**Comment by bogdando** [comment]
@tomassedovic 

>openshift_openstack_private_hostname_suffix: private
I expect it to create entries with the master-0.private or even master-0private hostname and openhsift.example.com zone.

Just to clarify, it should put 'master-0.private' entries (the shortname 'master-0' kept intact) living in a subdomain 'private' of the main (unchanged) openshift.example.com. So, FQDN would be master-0.private.openshift.example.com and nodes can resolve  master-0.private?

**Comment by tomassedovic** [comment]
I'm sorry @bogdando, I should have specified this more clearly. I'll try to do better next time.

This should not even treat it as a subdomain. Just a hostname/record suffix.

With `openshift_openstack_clusterid: "openshift"` and `openshift_openstack_public_dns_domain: "example.com"`:

If we set `openshift_openstack_private_hostname_suffix: "-private"`, the record we set would be `master-0-private`, the FQDN `master-0-private.openshift.example.com` and the zone: `openshift.example.com`.

Here's the nsupdate entry this should generate:

```yaml
nsupdate:
  record: "master-0-private"
  zone: "openshift.example.com"
...
```
The nodes resolve the FQDN (just like they do today), so:  `master-0-private.openshift.example.com`.


If we set `openshift_openstack_private_hostname_suffix: ".private"`, the record would be `master-0.private` (you can have dots in a record without changing the zone), the FQDN `master-0.private.openshift.example.com` and the zone would still be `openshift.example.com`.

The resulting nsupdate would look like this:

```yaml
nsupdate:
  record: "master-0.private"
  zone: "openshift.example.com"
...
```

But no matter what we put in `openshift_openstack_private_hostname_suffix` or `openshift_openstack_public_hostname_suffix`, the zone must not change.

**Comment by bogdando** [comment]
> This should not even treat it as a subdomain. Just a hostname/record suffix.

I'd like to not modify hostnames, yet using subdomains. Updated for that case. See the docs update to test please.

**Comment by tomassedovic** [comment]
Yes, the zone must not change. The feature we need is very simple: if you specify a suffix, that string gets concatenated in the `record` portion of the nsupdate module. That is all. No zone changes, no additional domain management.

And yeah, this shouldn't modify the hostnames. Just the nsupdate records.

**Comment by bogdando** [comment]
@tomassedovic [this](https://github.com/openshift/openshift-ansible/pull/6349/commits/72bff2569347ec5ee7e7796851672167bc2621ec#diff-097777c8f5cfbed859413fd60b6ee346R364) illustrates your case, yet using subdomains. I found this simpler to implement and bringing less changes into the code than manipulating the hostnames. I'd prefer to not touch hostnames at all. Let's please keep shortnames intact, if possible.

**Comment by bogdando** [comment]
@tomassedovic note, we already have role names suffixes that alter nova servers' names, we need no another altering suffix for this case IMO

File: roles/openshift_openstack/defaults/main.yml [hunk 1]
```diff
@@ -46,13 +46,15 @@ openshift_openstack_container_storage_setup:
...
-openshift_openstack_full_private_dns_domain: "{{ (openshift_openstack_private_hostname_suffix == '') | ternary(openshift_openstack_private_dns_domain, openshift_openstack_private_hostname_suffix + '.' + openshift_openstack_private_dns_domain) }}"
+openshift_openstack_public_subdomain: "{{ openshift_openstack_clusterid|trim + '.' + openshift_openstack_public_dns_domain }}"
+openshift_openstack_private_subdomain: "{{ openshift_openstack_clusterid|trim + '.' + openshift_openstack_private_dns_domain }}"
+openshift_openstack_full_public_dns_domain: "{{ (openshift_openstack_public_hostname_suffix|trim == '') | ternary(openshift_openstack_public_subdomain, openshift_openstack_public_hostname_suffix + '.' + openshift_openstack_public_subdomain) }}"
+openshift_openstack_full_private_dns_domain: "{{ (openshift_openstack_private_hostname_suffix|trim == '') | ternary(openshift_openstack_private_subdomain, openshift_openstack_private_hostname_suffix + '.' + openshift_openstack_private_subdomain) }}"
```
Comment (bogdando) [inline roles/openshift_openstack/defaults/main.yml]: @tomassedovic I think subdomains give more flexibility than just hostname suffixes. You cannot alter private domain fully with the latter it seems.

**Comment by bogdando** [comment]
Anyway, to make it with hostnames suffixes, let's think of integrating this into openshift_openstack_master_hostname et al as well ... @tomassedovic any ideas?

**Comment by tomassedovic** [comment]
This is what we need:

https://github.com/tomassedovic/openshift-ansible/commit/c053061525a7e823ebca2a8bdc0ea66e330a4aa9

Plus making sure this propagates to the dynamic inventory. That is all.

We can set the domain already by changing the cluster id and default subdomain. It is possible we might want to add support for two completely separate configurable domains at some point, but that should be a separate change.

**Comment by bogdando** [comment]
> We can set the domain already by changing the cluster id and default subdomain. It is possible we might want to add support for two completely separate configurable domains at some point, but that should be a separate change.

I see. I keep this PR for that future case then.

So I got it, it seems all we want is a single public DNS service capable to resolve
`foo.openshift.example.com` into a public/floating IP, while `foo-private.openshift.example.com` into a private IP, given a private hostname prefix '-private'. I'll take a look into that, thanks!

**Comment by bogdando** [comment]
@tomassedovic shall those suffixes prepend `openstack_<role name>_hostname` as well?

**Comment by bogdando** [comment]
So we decided to stick with the nsupdates only suffixes https://github.com/openshift/openshift-ansible/pull/6381 .
Abandoning this for now.

**Pull request closed without merging**